File: 110. Balanced Binary Tree/sol.py
class TreeNode:
    def __init__(self, val = 0, left = Optional[TreeNode], right = Optional[TreeNode]):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            if left == -1 or right == -1 or abs(left - right) > 1:
                return -1
            return 1 + max(left, right)

        depth = dfs(root)
        if depth == -1:
            return False
        return True

        

File: 110. Balanced Binary Tree/sol.go
package BalancedBinaryTree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func isBalanced(root *TreeNode) bool {

    var dfs func(node *TreeNode) int 
    dfs = func(node *TreeNode) int {
        if node == nil {
            return 0
        }
        left := dfs(node.Left)
        right := dfs(node.Right)
        if left < 0 || right < 0 || abs(left - right) > 1 {
            return -1
        }
        return 1 + max(left, right)
    }

    depth := dfs(root)

    return depth >= 0 
    
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}